### ISSUE TITLE
[BUG]: Hardware mipmapping is not working properly in any game using it.

### ISSUE THREAD

**Issue by gilderoylockhart01** [PCSX2/pcsx2#5347]
### Describe the Bug

When using the hardware render, any game which will makes use of mipmapping will never match the expected output.

At this point of the project, this is one issue that can represent an important compatibility factor for the hardware render, many games are right on the edge of being perfect if this is addressed.

+ this can improve the performance "by a lot" in some games (ex Harry Poter the chamber of secrets, Tom clancy Ghost Recon).

Using full or basic mipmapping options will not change anything in most games to this date (except Jak game but even still, there are problems).

### Reproduction Steps

Load any of the games mentioned in this issue and see the problem with mipmapping being off.

### Expected Behavior

The mipmapping handling should be accurate to what the software render or real console produces.

### PCSX2 Revision

v1.7.2242-1-g8db0b5282

### Operating System

Windows 11

### CPU

I9 9900K

### GPU

RTX 3070

### GS Settings

Basic or full mipmapping for the games that uses it.

### GS Window Screenshots

The bug in action:

Half life:

HW render with full mipmapping:
![hal hw mipmap](https://user-images.githubusercontent.com/90222976/149661220-65f234be-3cd7-4298-bef9-a85a6fb21d47.png)

SW render with mipmapping:

![hal sw mipmap](https://user-images.githubusercontent.com/90222976/149661235-bb429632-e64e-4436-95dc-bbad29b1641e.png)

SW render without mipmapping:

![hal sw nomipmap](https://user-images.githubusercontent.com/90222976/149661245-232cbe59-eac9-4405-8f0c-4c9e2acfda19.png)

Tom Clancy Ghost Recon:

HW render with full mipmapping:

![ghost hw mip](https://user-images.githubusercontent.com/90222976/149661261-c808ead3-dc95-43e2-9e0b-0fe8cec85216.png)

SW render with mipmapping:
![ghost sw mip](https://user-images.githubusercontent.com/90222976/149661268-4be3b010-0fd5-4249-9ac4-ffd03360f10e.png)

SW render without mipmapping:

![ghost sw nomip](https://user-images.githubusercontent.com/90222976/149661273-edbed981-7335-4869-998b-77de21d64885.png)

PaRappa the Rapper 2:

HW Render with full mipmapping:

![para HW mipmap](https://user-images.githubusercontent.com/90222976/149661288-380bc1ae-8246-4570-8172-9aa9275eed7d.png)

SW render with mipmapping:

![para sw mipmap](https://user-images.githubusercontent.com/90222976/149661298-594648dc-930d-4b39-9f9a-8d219efb59bb.png)

SW render without mipmapping:

![para sw nomipmap](https://user-images.githubusercontent.com/90222976/149661304-df0c35fc-5b04-4039-ac0d-a6fc31b12176.png)

### Logs & Dumps

GSdump of the various games affected by this: https://drive.google.com/file/d/1iFjf8Y-zDXoWYw67CGeJzLIWOZD9fehD/view?usp=sharing

**Comment by refractionpcsx2** [comment]
I will just point out that SW mipmapping, especially in 64bit doesn't work properly either.

Ratchet & Clank and Jak are pretty much fine with mipmapping and they require it to look correct, otherwise the screen would be full of garbage.  But when comparing HW to SW it's also better to compare native not upscaled, because you're improving the quality of the textures artficially. Also consider using Trilinear Filtering under the hacks section as that's kind of a component of Mipmapping also.

**Comment by gilderoylockhart01** [comment]
Yeah, I admit I kinda rushed those screenshots. But I tested with the 32bit PCSX2 and native resolution and it's the same, the main problem regarding Ratchet and Jak is that mipmapping on hw lacks that sort of "smooth" transition which is hard to show with a simple picture.

**Comment by refractionpcsx2** [comment]
That's what the Trilinear filtering does.

**Comment by gilderoylockhart01** [comment]
I did put it on ultra and for some reasons, Ratchet Jak and Ghost recon all do not improve with that option regardless of the render as you can see.

https://user-images.githubusercontent.com/90222976/149662571-60a74120-d5f8-4186-b676-12d2ee97a13a.mp4

**Comment by refractionpcsx2** [comment]
was that Mipmapping forced to Full and Trilinear on Ultra?

**Comment by gilderoylockhart01** [comment]
Actually I had to put the mipmapping on full, trilinear on Ultra "and" enabling preload frame data in order for the grass to properly work.

However there will still be issues with distant textures:

![image](https://user-images.githubusercontent.com/90222976/149662707-663f1239-81ad-43ad-a68b-c1fa3933fac1.png)

**Comment by refractionpcsx2** [comment]
ah yes the GPU Palette Conversion made have an effect on it also as it toggles between automatic and manual mipmaps.

**Comment by refractionpcsx2** [comment]
However if the Mipmapping is broken, GS Dumps would be appreciated.

**Comment by gilderoylockhart01** [comment]
I included the GSdump in the issue template. Actaully I will do a round up of my games to check if any others have problems with mipmapping.

**Comment by gilderoylockhart01** [comment]
Okay, I revamped the issue to more closely match what happens, a few games deffinitely have issues with it, but a lot less than what I expected thanks to refraction's sugestion.

Those are the GSdumps of the games that for sure have a problem with hw mipmapping: https://drive.google.com/file/d/17iLSSfB9uEoXDjhUHENZLTFKvudsYGgr/view?usp=sharing

As a side note, why do we have basic mipmapping/trilinear filtering despite them not actually being faster in any cases? Also, the end user might easily miss that trilinear filtering option, why not enabling that together with the mipmapping directly in the GSDX code?

**Comment by stenzek** [comment]
> As a side note, why do we have basic mipmapping/trilinear filtering despite them not actually being faster in any cases? Also, the end user might easily miss that trilinear filtering option, why not enabling that together with the mipmapping directly in the GSDX code?

It is faster. Basic generates mips on the GPU, full reads, expands and uploads them from GS memory.

**Comment by gilderoylockhart01** [comment]
Ahhh, so it is a sort of hardware or software implementation if I understood correctly?

**Comment by gilderoylockhart01** [comment]
Jak 2 can be included in that "meta":

![gs_20220116160502](https://user-images.githubusercontent.com/90222976/149665681-5dc111b8-18ae-4022-818e-264052b5e533.png)

[Jak 2 mipmapping issues hardware.zip](https://github.com/PCSX2/pcsx2/files/7877201/Jak.2.mipmapping.issues.hardware.zip)

**Comment by kazuya-555** [comment]
I notice the original R&C doesnt use trilinear if the option is set to automatic ? Not sure if normal when factoring the mipmaps. Although the mipmap levels seem obvious regardless of settings, although use slightly higher blending if I sent them and trilinear to max.

And when I watched video from the PS4 HD versions, the mipmap levels also seemed less obvious, unless they`re remade from the original versions.

**Comment by kazuya-555** [comment]
I completed R&C and the credits section uses half pictures unless using software mode.

**Comment by TheTechnician27** [comment]
[Scooby-Doo! Night of 100 Frights_SLUS-20349_20240623110121.zip](https://github.com/user-attachments/files/15945243/Scooby-Doo.Night.of.100.Frights_SLUS-20349_20240623110121.zip) Night of 100 Frights has a seemingly small SW mips issue here.

**Comment by christopheKohler** [comment]
![Image](https://github.com/user-attachments/assets/72ccda9b-e67f-4c00-ac1b-df0612cf5b9f)

Great emulator.
Dave mirra 2 have same mipmap issue. 
In software mode no problem, in hardware mode some textures have one wong mipmap. 
I can be seen in first level, just hanging around, very quick to test.
I've also tested replacment textures, same problems.
(and tested all possible settings, no setting fix it appart of "rendering=software").

**Comment by refractionpcsx2** [comment]
yes, we're aware of Dave Mirra, the game renders each of the mip levels to different targets, then tries to use them all together, and it just doesn't work. I don't think it'll be a five minute job to fix that, unfortunately.

**Comment by christopheKohler** [comment]
Thanks for the efforts. The game is fully playable, so that would be a cosmetic fix.
